File: src/scanner.ts
import { ALL_FORMATS, Input } from "mediabunny";
import { basename, dirname, extname, join, relative } from "std/path";
import { DenoFileSource } from "./media_source.ts";
import { toPosixPath } from "./path.ts";
import { parseMediaName } from "./parser.ts";
import type {
  Library,
  MediaFile,
  SidecarSubtitle,
  SubtitleTrack,
} from "./types.ts";

const VIDEO_EXTENSIONS = new Set([
  ".mp4",
  ".m4v",
  ".mov",
  ".mkv",
  ".webm",
  ".avi",
  ".wmv",
  ".flv",
  ".mpeg",
  ".mpg",
  ".ts",
  ".m2ts",
  ".ogv",
]);

const SUBTITLE_EXTENSIONS = new Set([".srt", ".vtt"]);
const METADATA_PROBE_CONCURRENCY = 3;

export async function scanLibrary(
  mediaDir: string,
  previous?: Library,
): Promise<Library> {
  const items: MediaFile[] = [];
  const paths: string[] = [];
  const subtitles: SidecarSubtitle[] = [];
  const previousByRelativePath = new Map(
    (previous?.items ?? []).map((item) => [item.relativePath, item]),
  );

  for await (const path of walkIfExists(mediaDir)) {
    paths.push(path);
  }

  for (const path of paths) {
    const extension = extname(path).toLowerCase();
    if (SUBTITLE_EXTENSIONS.has(extension)) {
      subtitles.push(buildSidecarSubtitle(mediaDir, path, extension));
    }
  }

  const videoPaths = paths.filter((path) =>
    VIDEO_EXTENSIONS.has(extname(path).toLowerCase())
  );
  const scanned = await mapWithConcurrency(
    videoPaths,
    METADATA_PROBE_CONCURRENCY,
    async (path) =>
      await scanVideo(path, mediaDir, previousByRelativePath, subtitles),
  );
  items.push(...scanned);

  items.sort((a, b) =>
    a.kind.localeCompare(b.kind) ||
    (a.showTitle ?? a.title).localeCompare(b.showTitle ?? b.title) ||
    (a.season ?? 0) - (b.season ?? 0) ||
    (a.episode ?? 0) - (b.episode ?? 0) ||
    a.title.localeCompare(b.title)
  );

  return {
    mediaDir,
    scannedAt: new Date().toISOString(),
    items,
  };
}

async function scanVideo(
  path: string,
  mediaDir: string,
  previousByRelativePath: Map<string, MediaFile>,
  subtitles: SidecarSubtitle[],
): Promise<MediaFile> {
  const extension = extname(path).toLowerCase();
  const info = await Deno.stat(path);
  const relativePath = toPosixPath(relative(mediaDir, path));
  const parsed = parseMediaName(relativePath);
  const modifiedAt = (info.mtime ?? new Date()).toISOString();
  const previousItem = previousByRelativePath.get(relativePath);
  const unchanged = previousItem?.size === info.size &&
    previousItem.modifiedAt === modifiedAt;
  const media: MediaFile = unchanged
    ? {
      ...previousItem,
      path,
      relativePath,
      size: info.size,
      modifiedAt,
      extension,
      ...parsed,
    }
    : {
      id: previousItem?.id ?? await stableId(relativePath),
      path,
      relativePath,
      size: info.size,
      modifiedAt,
      extension,
      ...parsed,
    };

  if (!unchanged) {
    Object.assign(media, await inspectMedia(path, info.size));
  }
  media.sidecarSubtitles = findSidecarSubtitles(media, subtitles);
  return media;
}

async function mapWithConcurrency<T, R>(
  values: T[],
  limit: number,
  map: (value: T) => Promise<R>,
): Promise<R[]> {
  const results = new Array<R>(values.length);
  let nextIndex = 0;
  const worker = async () => {
    while (true) {
      const index = nextIndex++;
      if (index >= values.length) return;
      results[index] = await map(values[index]);
    }
  };
  await Promise.all(
    Array.from({ length: Math.min(limit, values.length) }, worker),
  );
  return results;
}

async function* walkIfExists(dir: string): AsyncGenerator<string> {
  try {
    yield* walk(dir);
  } catch (error) {
    if (error instanceof Deno.errors.NotFound) {
      return;
    }

    throw error;
  }
}

async function* walk(dir: string): AsyncGenerator<string> {
  for await (const entry of Deno.readDir(dir)) {
    if (entry.name.startsWith(".")) {
      continue;
    }

    const path = join(dir, entry.name);
    if (entry.isDirectory) {
      yield* walk(path);
    } else if (entry.isFile) {
      yield path;
    }
  }
}

async function inspectMedia(
  path: string,
  size: number,
): Promise<Partial<MediaFile>> {
  const input = new Input({
    source: new DenoFileSource(path, size),
    formats: ALL_FORMATS,
  });

  try {
    const [durationSeconds, mimeType, videoTrack, audioTrack] = await Promise
      .all([
        input.getDurationFromMetadata(undefined, { skipLiveWait: true }).catch(
          () => null,
        ),
        input.getMimeType().catch(() => undefined),
        input.getPrimaryVideoTrack().catch(() => null),
        input.getPrimaryAudioTrack().catch(() => null),
      ]);
    let embeddedSubtitles: SubtitleTrack[] = await input.getTracks({
      filter: (track) => track.type === "subtitle",
    }).then((tracks) =>
      Promise.all(tracks.map(async (track) => ({
        codec: await track.getCodec().then((codec) => codec ?? undefined).catch(
          () => undefined,
        ),
        language: await track.getLanguageCode().then((language) =>
          language === "und" ? undefined : language
        ).catch(() => undefined),
        name: await track.getName().then((name) => name ?? undefined).catch(
          () => undefined,
        ),
        source: "mediabunny" as const,
        renderable: false,
      })))
    ).catch(() => []);
    embeddedSubtitles = mergeSubtitleTracks(
      embeddedSubtitles,
      await inspectEmbeddedSubtitlesWithFfprobe(path),
    );

    const [videoCodec, audioCodec, width, height] = await Promise.all([
      videoTrack?.getCodec().catch(() => null),
      audioTrack?.getCodec().catch(() => null),
      videoTrack?.getDisplayWidth().catch(() => undefined),
      videoTrack?.getDisplayHeight().catch(() => undefined),
    ]);

    return {
      durationSeconds: durationSeconds ?? undefined,
      mimeType,
      videoCodec: videoCodec ?? undefined,
      audioCodec: audioCodec ?? undefined,
      width,
      height,
      embeddedSubtitles,
    };
  } catch (error) {
    return {
      metadataError: error instanceof Error ? error.message : String(error),
    };
  } finally {
    input.dispose();
  }
}

async function inspectEmbeddedSubtitlesWithFfprobe(
  path: string,
): Promise<SubtitleTrack[]> {
  try {
    const command = new Deno.Command("ffprobe", {
      args: [
        "-v",
        "error",
        "-select_streams",
        "s",
        "-show_entries",
        "stream=index,codec_name,codec_type:stream_tags=language,title",
        "-of",
        "json",
        path,
      ],
      stdout: "piped",
      stderr: "null",
    });
    const output = await command.output();
    if (!output.success) {
      return [];
    }

    const raw = new TextDecoder().decode(output.stdout);
    const data = JSON.parse(raw) as {
      streams?: Array<{
        index?: number;
        codec_name?: string;
        tags?: {
          language?: string;
          title?: string;
        };
      }>;
    };

    return (data.streams ?? [])
      .filter((stream) => typeof stream.index === "number")
      .map((stream) => ({
        streamIndex: stream.index,
        codec: stream.codec_name,
        language: normalizeLanguage(stream.tags?.language),
        name: stream.tags?.title,
        source: "ffprobe",
        renderable: isRenderableEmbeddedSubtitle(stream.codec_name),
      }));
  } catch {
    return [];
  }
}

function mergeSubtitleTracks(
  primary: SubtitleTrack[],
  fallback: SubtitleTrack[],
): SubtitleTrack[] {
  const seen = new Set<string>();
  const merged: SubtitleTrack[] = [];

  for (const track of [...primary, ...fallback]) {
    const key = track.streamIndex !== undefined
      ? `stream:${track.streamIndex}`
      : `${track.codec ?? ""}:${track.language ?? ""}:${track.name ?? ""}`;
    if (seen.has(key)) {
      continue;
    }

    seen.add(key);
    merged.push(track);
  }

  return merged;
}

function isRenderableEmbeddedSubtitle(codec?: string): boolean {
  return ["subrip", "ass", "ssa", "webvtt"].includes(codec ?? "");
}

function normalizeLanguage(language?: string): string | undefined {
  return language && language !== "und" ? language : undefined;
}

function buildSidecarSubtitle(
  mediaDir: string,
  path: string,
  extension: string,
): SidecarSubtitle {
  const relativePath = toPosixPath(relative(mediaDir, path));
  const filename = basename(path, extension);
  const language = inferSubtitleLanguage(filename);

  return {
    path,
    relativePath,
    label: language ? language.toUpperCase() : basename(path),
    format: extension === ".vtt" ? "vtt" : "srt",
    language,
  };
}

function findSidecarSubtitles(
  media: MediaFile,
  subtitles: SidecarSubtitle[],
): SidecarSubtitle[] {
  const videoDir = dirname(media.path);
  const videoStem = basename(media.path, media.extension);

  return subtitles
    .filter((subtitle) => {
      if (dirname(subtitle.path) !== videoDir) {
        return false;
      }

      const subtitleStem = basename(subtitle.path, `.${subtitle.format}`);
      return subtitleStem === videoStem ||
        subtitleStem.startsWith(`${videoStem}.`);
    })
    .sort((a, b) => a.label.localeCompare(b.label));
}

function inferSubtitleLanguage(filename: string): string | undefined {
  const parts = filename.split(".");
  const maybeLanguage = parts.at(-1);

  if (maybeLanguage && /^[a-z]{2,3}(-[a-z]{2})?$/i.test(maybeLanguage)) {
    return maybeLanguage;
  }

  return undefined;
}

async function stableId(value: string): Promise<string> {
  const hash = await crypto.subtle.digest(
    "SHA-256",
    new TextEncoder().encode(value),
  );

  return Array.from(new Uint8Array(hash))
    .slice(0, 12)
    .map((byte) => byte.toString(16).padStart(2, "0"))
    .join("");
}

File: internal/cake/scanner.go
package cake

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"os"
	"os/exec"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
	"sync"
	"time"
)

var videoExtensions = map[string]bool{
	".mp4": true, ".m4v": true, ".mov": true, ".mkv": true,
	".webm": true, ".avi": true, ".wmv": true, ".flv": true,
	".mpeg": true, ".mpg": true, ".ts": true, ".m2ts": true,
	".ogv": true,
}

var subtitleExtensions = map[string]bool{".srt": true, ".vtt": true}

var mediaMIMEs = map[string]string{
	".mp4": "video/mp4", ".m4v": "video/mp4", ".mov": "video/quicktime",
	".mkv": "video/x-matroska", ".webm": "video/webm",
	".avi": "video/x-msvideo", ".wmv": "video/x-ms-wmv",
	".flv": "video/x-flv", ".mpeg": "video/mpeg", ".mpg": "video/mpeg",
	".ts": "video/mp2t", ".m2ts": "video/mp2t", ".ogv": "video/ogg",
}

func scanLibrary(mediaDir string, previous *Library) (*Library, error) {
	var videos []string
	var subtitles []SidecarSubtitle
	err := filepath.WalkDir(mediaDir, func(path string, entry os.DirEntry, err error) error {
		if err != nil {
			if errors.Is(err, os.ErrNotExist) && path == mediaDir {
				return filepath.SkipDir
			}
			return err
		}
		if entry.IsDir() {
			if path != mediaDir && strings.HasPrefix(entry.Name(), ".") {
				return filepath.SkipDir
			}
			return nil
		}
		if !entry.Type().IsRegular() || strings.HasPrefix(entry.Name(), ".") {
			return nil
		}
		ext := strings.ToLower(filepath.Ext(path))
		if videoExtensions[ext] {
			videos = append(videos, path)
		} else if subtitleExtensions[ext] {
			subtitles = append(subtitles, buildSidecar(mediaDir, path, ext))
		}
		return nil
	})
	if err != nil && !errors.Is(err, os.ErrNotExist) {
		return nil, err
	}
	previousByPath := make(map[string]MediaFile)
	if previous != nil {
		for _, item := range previous.Items {
			previousByPath[item.RelativePath] = item
		}
	}
	items := make([]MediaFile, len(videos))
	jobs := make(chan int)
	var wg sync.WaitGroup
	var firstErr error
	var errMu sync.Mutex
	workers := 3
	if len(videos) < workers {
		workers = len(videos)
	}
	for range workers {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for i := range jobs {
				item, e := scanVideo(videos[i], mediaDir, previousByPath, subtitles)
				if e != nil {
					errMu.Lock()
					if firstErr == nil {
						firstErr = e
					}
					errMu.Unlock()
				}
				items[i] = item
			}
		}()
	}
	for i := range videos {
		jobs <- i
	}
	close(jobs)
	wg.Wait()
	if firstErr != nil {
		return nil, firstErr
	}
	sort.Slice(items, func(i, j int) bool {
		a, b := items[i], items[j]
		if a.Kind != b.Kind {
			return a.Kind < b.Kind
		}
		at, bt := a.Title, b.Title
		if a.ShowTitle != "" {
			at = a.ShowTitle
		}
		if b.ShowTitle != "" {
			bt = b.ShowTitle
		}
		if at != bt {
			return at < bt
		}
		as, bs := intValue(a.Season), intValue(b.Season)
		if as != bs {
			return as < bs
		}
		ae, be := intValue(a.Episode), intValue(b.Episode)
		if ae != be {
			return ae < be
		}
		return a.Title < b.Title
	})
	return &Library{MediaDir: mediaDir, ScannedAt: isoTime(time.Now()), Items: items}, nil
}

func scanVideo(path, mediaDir string, previous map[string]MediaFile, subtitles []SidecarSubtitle) (MediaFile, error) {
	info, err := os.Stat(path)
	if err != nil {
		return MediaFile{}, err
	}
	rel, err := filepath.Rel(mediaDir, path)
	if err != nil {
		return MediaFile{}, err
	}
	rel = filepath.ToSlash(rel)
	modified := isoTime(info.ModTime())
	ext := strings.ToLower(filepath.Ext(path))
	parsed := parseMediaName(rel)
	old, exists := previous[rel]
	item := MediaFile{
		ID: stableID(rel), Path: path, RelativePath: rel,
		Size: info.Size(), ModifiedAt: modified, Extension: ext,
	}
	if exists && old.Size == info.Size() && old.ModifiedAt == modified {
		item = old
		item.Path = path
		item.RelativePath = rel
		item.Size = info.Size()
		item.ModifiedAt = modified
		item.Extension = ext
	} else {
		inspectMedia(path, &item)
	}
	item.Kind = parsed.Kind
	item.Title = parsed.Title
	item.ShowTitle = parsed.ShowTitle
	item.Season = parsed.Season
	item.Episode = parsed.Episode
	item.EpisodeTitle = parsed.EpisodeTitle
	item.Year = parsed.Year
	item.SidecarSubtitles = findSidecars(item, subtitles)
	return item, nil
}

type probeData struct {
	Format struct {
		Duration   string `json:"duration"`
		FormatName string `json:"format_name"`
	} `json:"format"`
	Streams []struct {
		Index     int    `json:"index"`
		CodecName string `json:"codec_name"`
		CodecType string `json:"codec_type"`
		Width     int    `json:"width"`
		Height    int    `json:"height"`
		Tags      struct {
			Language string `json:"language"`
			Title    string `json:"title"`
		} `json:"tags"`
	} `json:"streams"`
}

func inspectMedia(path string, item *MediaFile) {
	args := []string{
		"-v", "error",
		"-show_entries", "format=duration,format_name:stream=index,codec_name,codec_type,width,height:stream_tags=language,title",
		"-of", "json", path,
	}
	out, err := exec.Command("ffprobe", args...).Output()
	if err != nil {
		item.MetadataError = err.Error()
		return
	}
	var p probeData
	if err = json.Unmarshal(out, &p); err != nil {
		item.MetadataError = err.Error()
		return
	}
	if d, err := strconv.ParseFloat(p.Format.Duration, 64); err == nil {
		item.DurationSeconds = &d
	}
	item.MIMEType = mediaMIMEs[item.Extension]
	for _, s := range p.Streams {
		switch s.CodecType {
		case "video":
			if item.VideoCodec == "" {
				item.VideoCodec = s.CodecName
				if s.Width > 0 {
					v := s.Width
					item.Width = &v
				}
				if s.Height > 0 {
					v := s.Height
					item.Height = &v
				}
			}
		case "audio":
			if item.AudioCodec == "" {
				item.AudioCodec = s.CodecName
			}
		case "subtitle":
			idx := s.Index
			lang := s.Tags.Language
			if lang == "und" {
				lang = ""
			}
			item.EmbeddedSubtitles = append(item.EmbeddedSubtitles, SubtitleTrack{
				StreamIndex: &idx, Codec: s.CodecName,
				Language: lang, Name: s.Tags.Title, Source: "ffprobe",
				Renderable: oneOf(s.CodecName, "subrip", "ass", "ssa", "webvtt"),
			})
		}
	}
}

func buildSidecar(root, path, ext string) SidecarSubtitle {
	rel, _ := filepath.Rel(root, path)
	base := strings.TrimSuffix(filepath.Base(path), ext)
	parts := strings.Split(base, ".")
	lang := ""
	candidate := parts[len(parts)-1]
	if validLanguage(candidate) {
		lang = candidate
	}
	label := filepath.Base(path)
	if lang != "" {
		label = strings.ToUpper(lang)
	}
	format := "srt"
	if ext == ".vtt" {
		format = "vtt"
	}
	return SidecarSubtitle{
		Path: path, RelativePath: filepath.ToSlash(rel),
		Label: label, Format: format, Language: lang,
	}
}

func validLanguage(s string) bool {
	if len(s) < 2 || len(s) > 6 {
		return false
	}
	parts := strings.Split(s, "-")
	if len(parts) > 2 || len(parts[0]) < 2 || len(parts[0]) > 3 {
		return false
	}
	for _, part := range parts {
		for _, r := range part {
			if (r < 'a' || r > 'z') && (r < 'A' || r > 'Z') {
				return false
			}
		}
	}
	return len(parts) == 1 || len(parts[1]) == 2
}

func findSidecars(media MediaFile, subs []SidecarSubtitle) []SidecarSubtitle {
	var result []SidecarSubtitle
	dir := filepath.Dir(media.Path)
	stem := strings.TrimSuffix(filepath.Base(media.Path), media.Extension)
	for _, s := range subs {
		if filepath.Dir(s.Path) != dir {
			continue
		}
		subStem := strings.TrimSuffix(filepath.Base(s.Path), "."+s.Format)
		if subStem == stem || strings.HasPrefix(subStem, stem+".") {
			result = append(result, s)
		}
	}
	sort.Slice(result, func(i, j int) bool { return result[i].Label < result[j].Label })
	return result
}

func stableID(value string) string {
	sum := sha256.Sum256([]byte(value))
	return hex.EncodeToString(sum[:12])
}

func intValue(v *int) int {
	if v == nil {
		return 0
	}
	return *v
}

func isoTime(t time.Time) string {
	return t.UTC().Format("2006-01-02T15:04:05.000Z")
}
